contracts/utils/EmergencyBrake.sol

```diff
@@ -6,15 +6,17 @@ import "../access/AccessControl.sol";
 
 interface IEmergencyBrake {
     struct Permission {
-        address contact;
+        address contact; /// contract for which a user holds auth priviliges
```

Comment:
There is a fair amount of complexity in this contract that I think could be avoided if you refactor this struct to be:

```
struct Permission {
        address contact; /// contract for which the target holds auth privileges
        bytes4 signature;
}
```

Then you could have a Plan, which has a State and an array of Permissions. It doesn't need to store the `target`, as you don't need to iterate through Plans and can store all plans in a `mapping(address => Plan)` indexed by target.

Since we will want to iterate through the Permissions in a Plan, we can add to the Plan a `mapping(bytes32 => uint256) index` element for indexing the permissions.

Each time we add a Permission to a Plan, we add to the index the current length of the Permissions array in the Plan, using as index `(bytes32(_permission.signature) << 20 | bytes32(_permission.contact))`. Maybe you can have `permissionToId` and `permissionFromId` functions.

Each time we remove a Permission from a Plan, instead of rewriting a chunk of the array, we overwrite the permission removed with the last permission in the plan, update the index for the last permission in the plan, and pop the (now repeated) last permission. We zero the index of the removed permission, of course.

Reply:
> Then you could have a Plan, which has a State and an array of Permissions. It doesn't need to store the target, as you don't need to iterate through Plans and can store all plans in a mapping(address => Plan) indexed by target.

I used the original version of the base so a lot of the structure is still based on that. Changing it to the way you've described makes sense based on the new implementation. 

Since we will want to iterate through the Permissions in a Plan, we can add to the Plan a mapping(bytes32 => uint256) index element for indexing the permissions. 

> Since we will want to iterate through the Permissions in a Plan, we can add to the Plan a mapping(bytes32 => uint256) index element for indexing the permissions. 

So keep the array for iteration, and index the Permissions within each Plan. Right now they're indexed in a contract level mapping but realizing that creates the potential for collision 

> Each time we add a Permission to a Plan, we add to the index the current length of the Permissions array in the Plan, using as index (bytes32(_permission.signature) << 20 | bytes32(_permission.contact)). Maybe you can have permissionToId and permissionFromId functions.

Sounds like a good quality of life feature for when plans need to be changed or updated

>Each time we remove a Permission from a Plan, instead of rewriting a chunk of the array, we overwrite the permission removed with the last permission in the plan, update the index for the last permission in the plan, and pop the (now repeated) last permission. We zero the index of the removed permission, of course.

This make sense since we need to preserve the order for index mapping, but it doesn't have to be in the exact same order as before.